aptos-move/move-examples/data_structures/sources/smart_tree.move

```diff
@@ -0,0 +1,971 @@
+/// Type of large-scale search trees.
+///
+/// It internally uses BTree to organize the search tree data structure for keys. Comparing with
+/// other common search trees like AVL or Red-black tree, a BTree node has more children, and packs
+/// more metadata into one node, which is more disk friendly (and gas friendly).
+
+module aptos_std::smart_tree {
+    use aptos_std::bucket_table::{Self, BucketTable};
+    use std::option::{Self, Option};
+    use std::vector;
+
+    // Internal errors.
+    const E_INTERNAL: u64 = 0;
+    // The tree is not empty, and cannot be destroyed.
+    const E_TREE_NOT_EMPTY: u64 = 1;
+    // The tree is too big for insertion.
+    const E_TREE_TOO_BIG: u64 = 2;
+    // The provided parameter is invalid.
+    const E_INVALID_PARAMETER: u64 = 3;
+
+    const NULL_INDEX: u64 = 0;
```

Comment:
So, `0` can be a key because it's reserved for NULL?

Reply:
indices are a different concept than keys

index is the key of the table that stores nodes.